{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Boilerplate and imports\n",
    "\n",
    "import csv, glob, json, os, math, numbers, pandas, re, scipy, scipy.sparse, shutil\n",
    "import subprocess, sys, thread, threading, time, urllib2\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "def exec_ipynb(filename_or_url):\n",
    "    nb = (urllib2.urlopen(filename_or_url) if re.match(r'https?:', filename_or_url) else open(filename_or_url)).read()\n",
    "    jsonNb = json.loads(nb)\n",
    "    #check for the modified formatting of Jupyter Notebook v4\n",
    "    if(jsonNb['nbformat'] == 4):\n",
    "        exec '\\n'.join([''.join(cell['source']) for cell in jsonNb['cells'] if cell['cell_type'] == 'code']) in globals()\n",
    "    else:\n",
    "        exec '\\n'.join([''.join(cell['input']) for cell in jsonNb['worksheets'][0]['cells'] if cell['cell_type'] == 'code']) in globals()\n",
    "\n",
    "exec_ipynb('timelapse-utilities.ipynb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#notebook-container { margin-left:-14px; width:calc(100% + 27px) !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Wide display\n",
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>#notebook-container { margin-left:-14px; width:calc(100% + 27px) !important; }</style>\"))\n",
    "pd.set_option('display.max_columns', 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Per-school Common Core Data (CCD) from NCES\n",
    "# https://nces.ed.gov/ccd/pubschuniv.asp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Install r\n",
    "# sudo apt-get install r-base sudo libhdf5-dev\n",
    "# sudo R\n",
    "# install.packages(\"haven\")\n",
    "\n",
    "# library(haven)\n",
    "# write.csv()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccd_download_dir = 'allegheny_county/schools/nces'\n",
    "\n",
    "def uniqpath():\n",
    "    return '%s-%s' % (thread.get_ident(), os.getpid())\n",
    "\n",
    "# Very slow conversion ... but pandas read_sas is failing a lot\n",
    "def convert_sas_to_h5(sas):\n",
    "    h5 = os.path.splitext(sas)[0] + '.h5'\n",
    "    if os.path.exists(h5):\n",
    "        sys.stdout.write('%s already exists\\n' % h5)\n",
    "    else:\n",
    "        h5tmp = h5 + uniqpath() + \".tmp.h5\"\n",
    "        cmd = 'library(haven); library(rhdf5); h5createFile(\"%s\"); h5write(read_sas(\"%s\"), file=\"%s\", name=\"data\")' % (h5tmp, sas, h5tmp)\n",
    "        sys.stdout.write('Converting %s to %s\\n' % (sas, h5))\n",
    "        sys.stdout.write('%s\\n' % cmd)\n",
    "        !R_DEFAULT_PACKAGES= Rscript -e '$cmd'\n",
    "        os.rename(h5tmp, h5)\n",
    "    return h5\n",
    "\n",
    "# Reads src_url into table, via intermediate caches of downloading original source and converting to local HDF5\n",
    "def read_ccd_src(src_url):\n",
    "    dest = ccd_download_dir + '/' + os.path.basename(src_url)\n",
    "    download_file(src_url, dest)\n",
    "    unzip_file(dest)\n",
    "    sasfile = glob.glob(os.path.splitext(dest)[0] + '/*.sas7bdat')[0]\n",
    "    \n",
    "    hdffile = convert_sas_to_h5(sasfile)\n",
    "\n",
    "    sys.stdout.write('Reading %s\\n' % hdffile)\n",
    "    return pd.read_hdf(hdffile)\n",
    "    \n",
    "def ccd_from_year(year):\n",
    "    src = ccd_src_from_year(year)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_cols(a, b):\n",
    "    ret = a.copy()\n",
    "    use_b = a.isna() & ~b.isna()\n",
    "    ret[use_b] = b[use_b]\n",
    "    return ret\n",
    "\n",
    "def merge_ccd_dataframes(frames, join_col):\n",
    "    m = frames[0]\n",
    "    for rhs in frames[1:]:\n",
    "        cols_to_merge = sorted(set(m.columns) & set(rhs.columns) - set([join_col]))\n",
    "        m = pd.merge(m, rhs, on=join_col, how='outer', validate='1:1', suffixes=('_xxx', '_yyy'))\n",
    "        for col in cols_to_merge:\n",
    "            m[col] = merge_cols(m[col + '_xxx'], m[col + '_yyy'])\n",
    "            m.drop(columns = [col + '_xxx', col + '_yyy'], inplace=True)\n",
    "    return m\n",
    "\n",
    "def read_ccd(year):\n",
    "    srcs = ccd_src_from_year(year)\n",
    "\n",
    "def read_and_merge_ccd_srcs(srcs, horizontal=False):\n",
    "    if horizontal:\n",
    "        [read_ccd_src(src) for src in srcs]\n",
    "    else:\n",
    "        return merge_ccd_dataframes([read_ccd_src(src) for src in srcs], 'NCESSCH')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_2017(): return read_ccd_src('https://nces.ed.gov/ccd/Data/zip/ccd_sch_029_1718_w_0a_03302018_sas.zip')\n",
    "\n",
    "def compute_2016(): return read_ccd_src('https://nces.ed.gov/ccd/Data/zip/ccd_sch_029_1617_w_0e_050317_sas.zip')\n",
    "\n",
    "def compute_2015(): return read_and_merge_ccd_srcs([\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_029_1516_w_2a_011717_sas.zip',\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_052_1516_w_2a_011717_sas.zip',\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_059_1516_w_2a_011717_sas.zip',\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_129_1516_w_2a_011717_sas.zip',\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_033_1516_w_2a_011717_sas.zip'])\n",
    "\n",
    "def compute_2014(): return read_and_merge_ccd_srcs([\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_029_1415_w_0216161a_sas.zip',\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_052_1415_w_0216161a_sas.zip',\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_059_1415_w_0216161a_sas.zip',\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_129_1415_w_0216161a_sas.zip',\n",
    "    'https://nces.ed.gov/ccd/Data/zip/ccd_sch_033_1415_w_0216161a_sas.zip'])\n",
    "\n",
    "def compute_2013(): return read_ccd_src('https://nces.ed.gov/ccd/Data/zip/sc132a_sas.zip')\n",
    "\n",
    "def compute_2012(): return read_ccd_src('https://nces.ed.gov/ccd/Data/zip/sc122a_sas.zip')\n",
    "\n",
    "def compute_2011(): return read_ccd_src('https://nces.ed.gov/ccd/Data/zip/sc111a_supp_sas.zip')\n",
    "\n",
    "def compute_2010(): return read_ccd_src('https://nces.ed.gov/ccd/Data/zip/sc102a_sas7bdat.zip')\n",
    "\n",
    "def compute_2009(): return read_ccd_src('https://nces.ed.gov/ccd/data/zip/sc092a_sas.zip')\n",
    "\n",
    "def compute_2008(): return read_ccd_src('https://nces.ed.gov/ccd/data/zip/sc081b_sas.zip')\n",
    "\n",
    "def compute_2007(): return read_ccd_src('https://nces.ed.gov/ccd/data/zip/sc071b_sas.zip')\n",
    "\n",
    "def compute_2006(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc061cai_sas.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc061ckn_sas.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc061cow_sas.zip'],\n",
    "                                         horizontal=True)\n",
    "def compute_2005(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc051aai_sas.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc051akn_sas.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc051aow_sas.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "def compute_2004(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc041bai_sas.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc041bkn_sas.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc041bow_sas.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "def compute_2003(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc031aai_sas.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc031akn_sas.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc031aow_sas.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "def compute_2002(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc021aai_sd2.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc021akn_sd2.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc021aow_sd2.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "def compute_2001(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc011aai.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc011akn.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc011aow.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "def compute_2000(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc001aai.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc001akn.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc001aow.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "def compute_1999(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc991bai.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc991bkn.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc991bow.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "def compute_1998(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/sc981cai.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc981ckn.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/sc981cow.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "def compute_1997(): return read_ccd_srcs(['https://nces.ed.gov/ccd/data/zip/psu97ai.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/psu97kn.zip',\n",
    "                                          'https://nces.ed.gov/ccd/data/zip/psu97ow.zip'],\n",
    "                                         horizontal=True)\n",
    "                                         \n",
    "\n",
    "    \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "allegheny_county/schools/nces/2017-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2016-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2015-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/ccd_sch_029_1415_w_0216161a_sas.zip already downloaded\n",
      "allegheny_county/schools/nces/ccd_sch_029_1415_w_0216161a_sas.zip already unzipped\n",
      "allegheny_county/schools/nces/ccd_sch_029_1415_w_0216161a_sas/ccd_sch_029_1415_w_0216161a.h5 already exists\n",
      "Reading allegheny_county/schools/nces/ccd_sch_029_1415_w_0216161a_sas/ccd_sch_029_1415_w_0216161a.h5\n",
      "allegheny_county/schools/nces/2017-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2016-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2015-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/ccd_sch_029_1415_w_0216161a_sas.zip already downloaded\n",
      "allegheny_county/schools/nces/ccd_sch_029_1415_w_0216161a_sas.zip already unzipped\n",
      "allegheny_county/schools/nces/ccd_sch_029_1415_w_0216161a_sas/ccd_sch_029_1415_w_0216161a.h5 already exists\n",
      "Reading allegheny_county/schools/nces/ccd_sch_029_1415_w_0216161a_sas/ccd_sch_029_1415_w_0216161a.h5\n",
      "allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas.zip already downloaded\n",
      "allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas.zip already unzipped\n",
      "allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas/ccd_sch_052_1415_w_0216161a.h5 already exists\n",
      "Reading allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas/ccd_sch_052_1415_w_0216161a.h5\n",
      "allegheny_county/schools/nces/2012-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2011-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2013-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2010-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2009-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2008-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2007-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas.zip already downloaded\n",
      "allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas.zip already unzipped\n",
      "allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas/ccd_sch_052_1415_w_0216161a.h5 already exists\n",
      "Reading allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas/ccd_sch_052_1415_w_0216161a.h5\n",
      "2014 failed\n",
      "allegheny_county/schools/nces/2013-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2012-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2011-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2010-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2009-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2008-ccd-computed.h5 already exists, skipping\n",
      "allegheny_county/schools/nces/2007-ccd-computed.h5 already exists, skipping\n",
      "2006 failed\n",
      "2005 failed\n",
      "2004 failed\n",
      "2003 failed\n",
      "2002 failed\n",
      "2001 failed\n",
      "2000 failed\n",
      "1999 failed\n",
      "1998 failed\n",
      "1997 failed\n",
      "1996 failed\n",
      "1995 failed\n",
      "1994 failed\n",
      "1993 failed\n",
      "1992 failed\n",
      "1991 failed\n",
      "1990 failed\n",
      "1989 failed\n",
      "1988 failed\n",
      "1987 failed\n",
      "1986 failed\n",
      "1985 failed\n",
      "1984 failed\n",
      "1983 failed\n",
      "1982 failed\n",
      "1981 failed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: global name 'read_ccd_srcs' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1996' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1995' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1994' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1993' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1992' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1991' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1990' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1989' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1988' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1987' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1986' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1985' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1984' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1983' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1982' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "NameError: name 'compute_1981' is not defined\n",
      "Exception caught in SimpleProcessPoolExecutor.shutdown.  Continuing until all are finished.\n",
      "Exception follows:\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 215, in shutdown\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 455, in result\n",
      "    return self.__get_result()\n",
      "  File \"/home/rsargent/anaconda2/lib/python2.7/site-packages/concurrent/futures/_base.py\", line 414, in __get_result\n",
      "    raise exception_type, self._exception, self._traceback\n",
      "ValueError: No dataset in HDF5 file.\n"
     ]
    },
    {
     "ename": "Exception",
     "evalue": "SimpleProcessPoolExecutor failed: 27 of 37 raised exception",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mException\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-4cc97b23f415>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m         \u001b[0;32mprint\u001b[0m \u001b[0;34m'%d failed'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0myear\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m \u001b[0mpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshutdown\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<string>\u001b[0m in \u001b[0;36mshutdown\u001b[0;34m(self)\u001b[0m\n",
      "\u001b[0;31mException\u001b[0m: SimpleProcessPoolExecutor failed: 27 of 37 raised exception"
     ]
    }
   ],
   "source": [
    "def compute_if_needed(year):\n",
    "    filename = ccd_download_dir + ('/%d-ccd-computed.h5' % year)\n",
    "    if os.path.exists(filename):\n",
    "        print '%s already exists, skipping' % filename\n",
    "        return\n",
    "    ccd = eval('compute_%d()' % year)\n",
    "    tmp = filename + uniqpath() + '.h5'\n",
    "    ccd.to_hdf(tmp, 'data')\n",
    "    os.rename(tmp, filename)\n",
    "    print 'Created %s' % filename\n",
    "\n",
    "pool = SimpleProcessPoolExecutor(4)\n",
    "\n",
    "for year in range(2017, 1980, -1):\n",
    "    try:\n",
    "        pool.submit(compute_if_needed, year)\n",
    "        compute_if_needed(year)\n",
    "    except:\n",
    "        print '%d failed' % year\n",
    "        \n",
    "pool.shutdown()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!file allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas/*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_sas('allegheny_county/schools/nces/ccd_sch_052_1415_w_0216161a_sas/ccd_sch_052_1415_w_0216161a.sas7bdat')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_hdf(\"bar.h5\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
